### ISSUE TITLE
[Discussion] XML UI

### ISSUE THREAD

**Issue by dustinwloring1988** [TeamPorcupine/ProjectPorcupine#537]
I was trying to start on a XML style UI and was wondering what everyone thought about it. This will allow for easy editing with out touching the _SCENE_ file. Also just like the whole concept of this project it will be easy to mod after the project is built.
All input is welcome good, bad or indifferent.

**Comment by bjubes** [comment]
The only concern is that there is A LOT of stuff unity takes care of when making UI. Nine sprite splice, anchors, parent and child anchors and positions, etc....

If this is changed to XML, the work has to be done manually, which would take much longer. Its the same problem with making Spritesheets in XML, its a pain to type out a file of each wall and specify each x,y,width,height,ppu of each variation. This will be nothing compared to the tedious nature of making an entire GUI, which currently is  23326 lines in unity's scene file

**Comment by fre-ber** [comment]
@bjubes I think that you are correct about the UI it would be quite tricky to get a flexible UI this way. As for extracting sprite sheets from the meta-files, I made a little script for that, posted in the comments of episode 64 of the tutorial video series. Perhaps it can be useful. I haven't tested it with the latest version of Unity, though. But if something changed, it shouldn't be too hard to modify the script to make it work again. Perhaps we should add a folder for external tools like this in the repo.

**Comment by ttyz** [comment]
I think the best solution is to make some reusable components, like dialog, buttons, fields, which will be stored as prefabs, or preferably generated on the fly, and only reference those in the xml files:
```xml
<Dialog title="Save Game">
    <FileList name="fileList" />
</Dialog>
``` 

Trying to replicate a UI system with the same level of flexibility Unity has is too big of a task I think.

**Comment by vogonistic** [comment]
One big reason to have something that is entirely covered outside of the scene (but might be assets, idk) is modding. If we want them to add buttons, dialogs, move things around or generally do really cool things we haven't thought about, we need a system that is flexible enough for that.

We should implement our own UI as a mod, IMO.

**Comment by b4ux1t3** [comment]
Would it not be possible to have some kind of parser for the XML files, but that still uses Unity's UI controls behind the scenes? If I'm not mistaken, that's how Kerbal Space Program does it.

**Comment by vogonistic** [comment]
@b4ux1t3 I think that was what @ttyz was suggesting.

**Comment by dustinwloring1988** [comment]
Yes that is what I had going. I have prefabs made of the common game objects like buttons and panels with certin scripts attached like localization. Only the basic options like position anchors and OnClick are available to made by xml at lest to start. You can also call any prefab from a predefind list of like the dialog boxs and the langage dropdown.

**Comment by vogonistic** [comment]
This is a naive question because I'm not really that familiar with the Unity UI, but what is the big benefit of using prefabs as opposed to creating them with code?

**Comment by dustinwloring1988** [comment]
I plled the XML UI PR since I just started call prefabs from a script for now. Here is the PR I posted with code driven Dialog Boxes https://github.com/TeamPorcupine/ProjectPorcupine/pull/659

**Comment by TomMalbran** [comment]
For the bottom Menu we could have a ```MenuManager``` that will allow to register ```MenuItems``` each with a name, position, onClick callback, and the manager will be responsible for drawing the menu correctly and calling the callbacks.

**Comment by sboigelot** [comment]
as #659 is merged, should we close this issue? Most of the UI is code driven now.

**Comment by TomMalbran** [comment]
The UI uses Prefabs. If we want to go for a real code driven UI we need to do a lot more work.
